### ISSUE TITLE
[Bug]: PHP Parse Error in Laravel Extension on DDEV Environment with WSL2

### ISSUE THREAD

**Issue by krudi** [laravel/vs-code-extension#85]
### Extension Version

0.1.12

### PHP Binary

Docker

### Operating System

Windows

### What happened?

When I activate the Laravel Extension in VS Code while using a DDEV-based Docker environment on WSL2, I get a PHP Parse Error stating there's an `Unclosed (` in the command-line code executed by the extension. Even though the Laravel project is running correctly in the environment, the extension doesn’t initialize. Do I need to set up any special configuration for this setup to make it work?

I wanted to mention that I have an alias for the PHP command, which executes everything in the PHP container. The alias is as follows:

```
path=$(printf '%s\n' "${PWD##*/}")
command="ddev php "$@""
$command
```

### Mimimal Code Sample

```php
024-12-18 20:46:19.835 [error] Error:

PHP Parse error:  Unclosed '(' in Command line code on line 1
[31mFailed to run php -r error_reporting(E_ERROR | E_PARSE); define('LARAVEL_START', microtime(true)); require_once '/home/user/project/vendor/autoload.php'; $app = require_once '/home/user/project/bootstrap/app.php'; class VsCodeLaravel extends \Illuminate\Support\ServiceProvider { public function register() { } public function boot() { if (method_exists($this->app['log'], 'setHandlers')) { $this->app['log']->setHandlers([new \Monolog\Handler\ProcessHandler()]); } } } $app->register(new VsCodeLaravel($app)); $kernel = $app->make(Illuminate\Contracts\Console\Kernel::class); $kernel->bootstrap(); echo '__VSCODE_LARAVEL_START_OUTPUT__'; echo json_encode([ [ 'key' => 'base_path', 'path' => base_path(), ], [ 'key' => 'resource_path', 'path' => resource_path(), ], [ 'key' => 'config_path', 'path' => config_path(), ], [ 'key' => 'app_path', 'path' => app_path(), ], [ 'key' => 'database_path', 'path' => database_path(), ], [ 'key' => 'lang_path', 'path' => lang_path(), ], [ 'key' => 'public_path', 'path' => public_path(), ], [ 'key' => 'storage_path', 'path' => storage_path(), ], ]); ; echo '__VSCODE_LARAVEL_END_OUTPUT__'; exit(0);: exit status 255[0m
```

**Comment by joetannenbaum** [comment]
We just released v0.1.13 that should have better Windows support, can you verify that this is still an issue after upgrading?

**Comment by krudi** [comment]
> We just released v0.1.13 that should have better Windows support, can you verify that this is still an issue after upgrading?

Now I'm getting a slightly different error:

```
Error:
 Middlewares

Could not open input file: {code}
[31mFailed to run php {code} /home/user/project/vendor/_laravel_ide/discover-cc27311d80e3a317bc2fdb63b0e45636.php: exit status 1[0m

php "{code}" /home/user/project/vendor/_laravel_ide/discover-cc27311d80e3a317bc2fdb63b0e45636.php
```

**Comment by mysticseagull** [comment]
I am also receiving a similar error. any fixes for these would be great!

Error:

Could not open input file: /home/[REDACTED]/basil-website/vendor/_laravel_ide/discover-ccd6cc1aa7eff2077ce4a32717f83d56.php

php /home/[REDACTED]/basil-website/vendor/_laravel_ide/discover-ccd6cc1aa7eff2077ce4a32717f83d56.php

**Comment by joetannenbaum** [comment]
Are there spaces in your path? If so, we can close this issue in favor of #19

**Comment by krudi** [comment]
> Are there spaces in your path? If so, we can close this issue in favor of #19

No, my current project path is `/home/user/phpext` and there are no spaces in the path.

**Comment by joetannenbaum** [comment]
Wanted to check back in and see if this is still happening with v0.1.16, let me know if there is any improvement. Also, I noticed that you have `{code}` in your custom PHP command, feel free to remove that and see if it works.

**Comment by krudi** [comment]
I have now dug deeper into the problem and noticed that the command:

```
php /var/www/html/vendor/_laravel_ide/discover-e6ed03e84a7e56c2b0f65d30270ea14c.php
```

works correctly, and I get the following output:

`__VSCODE_LARAVEL_START_OUTPUT__{‘page_paths’:[]}__VSCODE_LARAVEL_END_OUTPUT__%`

However, when I run the same command with an absolute path that leads directly to the file outside the Docker container, such as:

`php /home/user/project/vendor/_laravel_ide/discover-e6ed03e84a7e56c2b0f65d30270ea14c.php`

I get the following error:

`Failed to execute command php /home/user/project/vendor/_laravel_ide/discover-e6ed03e84a7e56c2b0f65d30270ea14c.php: exit status 1`

This is because the path `/home/user/project/` refers to the local file system, and when running PHP outside the Docker container, it cannot access the file located inside the containerized environment (`/var/www/html/`).

**Comment by hemilioaraujo** [comment]
> I have now dug deeper into the problem and noticed that the command:
> 
> ```
> php /var/www/html/vendor/_laravel_ide/discover-e6ed03e84a7e56c2b0f65d30270ea14c.php
> ```
> 
> works correctly, and I get the following output:
> 
> `__VSCODE_LARAVEL_START_OUTPUT__{‘page_paths’:[]}__VSCODE_LARAVEL_END_OUTPUT__%`
> 
> However, when I run the same command with an absolute path that leads directly to the file outside the Docker container, such as:
> 
> `php /home/user/project/vendor/_laravel_ide/discover-e6ed03e84a7e56c2b0f65d30270ea14c.php`
> 
> I get the following error:
> 
> `Failed to execute command php /home/user/project/vendor/_laravel_ide/discover-e6ed03e84a7e56c2b0f65d30270ea14c.php: exit status 1`
> 
> This is because the path `/home/user/project/` refers to the local file system, and when running PHP outside the Docker container, it cannot access the file located inside the containerized environment (`/var/www/html/`).

Same situation here.
I set `"Laravel.phpCommand": "docker exec -w /var/www/html -u 1000:1000 workout-plan-laravel.test-1 php"`
I'm using windows, wsl and docker.

**Comment by krudi** [comment]
> > I have now dug deeper into the problem and noticed that the command:
> > ```
> > php /var/www/html/vendor/_laravel_ide/discover-e6ed03e84a7e56c2b0f65d30270ea14c.php
> > ```
> > 
> > 
> >     
> >   
> > works correctly, and I get the following output:
> > `__VSCODE_LARAVEL_START_OUTPUT__{‘page_paths’:[]}__VSCODE_LARAVEL_END_OUTPUT__%`
> > However, when I run the same command with an absolute path that leads directly to the file outside the Docker container, such as:
> > `php /home/user/project/vendor/_laravel_ide/discover-e6ed03e84a7e56c2b0f65d30270ea14c.php`
> > I get the following error:
> > `Failed to execute command php /home/user/project/vendor/_laravel_ide/discover-e6ed03e84a7e56c2b0f65d30270ea14c.php: exit status 1`
> > This is because the path `/home/user/project/` refers to the local file system, and when running PHP outside the Docker container, it cannot access the file located inside the containerized environment (`/var/www/html/`).
> 
> Same situation here. I set `"Laravel.phpCommand": "docker exec -w /var/www/html -u 1000:1000 workout-plan-laravel.test-1 php"` I'm using windows, wsl and docker.

Thank you for your reply. The setting with `phpCommand` that you mentioned does seem to be working for me as well. However, I’ve run into another issue where the Laravel extension still cannot find the artisan file. Is this maybe related to user/file permissions?

My current config

```
"Laravel.basePath": "/var/www/html",
"Laravel.phpCommand": "docker exec -w /var/www/html -u 1002:1002 ddev-project-web php"
```

Logs:

```
2025-01-23 12:20:47.168 [info] Activating Laravel Extension...
2025-01-23 12:20:47.168 [info] Not activating Laravel Extension because no artisan file found
2025-01-23 12:20:47.168 [info] Not activating Laravel Extension because "shouldActivate" returned false
```

File inside the Docker Container:

```
user@docker-container:/var/www/html$ ls -la artisan
-rwxr-xr-x 1 user user 350 Nov 24 18:09 artisan
```

**Comment by TonsiTT** [comment]
Have a ddev installation on win 11 with docker inside wsl (no docker desktop on windows) and getting same errors.
On default installation, I get errors 
```
 Could not open input file: /home/username/project_name/vendor/_laravel_ide/discover-1b81dad6ceb4b12f1398b0e8d88063fe.php

2025-02-20 18:44:47.736 [info] Checking local PHP installation: php -r 'echo PHP_BINARY;'
2025-02-20 18:44:47.736 [info] Falling back to system PHP installation
2025-02-20 18:44:47.737 [error] Error:
```

Setting base path as `"Laravel.basePath": "/var/www/html",` gives these errors instead
```
2025-01-23 12:20:47.168 [info] Activating Laravel Extension...
2025-01-23 12:20:47.168 [info] Not activating Laravel Extension because no artisan file found
2025-01-23 12:20:47.168 [info] Not activating Laravel Extension because "shouldActivate" returned false
```

**Comment by rfay** [comment]
DDEV maintainer here... It's hard to get VS Code to use the `php` inside the container for execution. 

The two good ways to do this that I know of are:

1. Use the "remote explorer" extension, so that you're working right inside the DDEV web container.
2. Use a fake `php`, for example, a /usr/local/bin/php with contents

```bash
#!/bin/bash
ddev php "$@"
```
And configure VS code to use php.executablePath=/usr/local/bin/php

**Comment by rfay** [comment]
> maybe it's because I also have PHP installed inside wsl and the extension is using that one ? Is there a side effect I didn't notice

Yes, you are probably using the `php` inside WSL, which may have other side-effects, especially if a different version or other different behavior.

**Comment by TonsiTT** [comment]
> DDEV maintainer here... It's hard to get VS Code to use the `php` inside the container for execution.
> 
> The two good ways to do this that I know of are:
> 
>     1. Use the "remote explorer" extension, so that you're working right inside the DDEV web container.
> 
>     2. Use a fake `php`, for example, a /usr/local/bin/php with contents
> 
> 
> #!/bin/bash
> ddev php "$@"
> 
> And configure VS code to use php.executablePath=/usr/local/bin/php

In the case of this extension, it doesn't work with alias or not.

Are dev containers via remote explorer supported/recommended by ddev?

**Comment by rfay** [comment]
I'm not a great VS Code user but have experimented with devcontainers/remote explorer and it definitely works, but it's not the normal documented way.  @ultimike has given lots of presentations on the strategy, like https://www.youtube.com/watch?v=Nn_L6N4ZQqo

This is a place where VS Code is behind JetBrains.

**Comment by marcelwirtz** [comment]
I managed to get it working for my laravel projects with the following solution.

First i created a executeable file `/usr/local/bin/php.laravel`
```
path=$1
path=${path#*/vendor}
path="/var/www/html/vendor$path"
command="ddev exec php "$path""
$command
```
and in the extension setting for the Laravel Plugin i set the PHP Command to `php.laravel {code}`.

It is still an ugly fix but should work for most of my projects.

**Comment by szilvesztercsab** [comment]
This fake PHP executable:

```bash
#!/usr/bin/env bash

# ddev php "$@"

if [[ $1 == *"vendor"* ]]; then
    path=$1
    path=${path#*/}
    path="/var/www/html/vendor$path"
    command="ddev php $path ${@:2}"
else
    command="ddev php $@"
fi
$command
``` 

And these settings worked for me:
```json
{
    "php.debug.executablePath": ".vscode/php",
    "php.validate.executablePath": ".vscode/php",
    "Laravel.phpCommand": ".vscode/php \"{code}\"",
    "LaravelExtraIntellisense.phpCommand": ".vscode/php -r \"{code}\"",
    "LaravelExtraIntellisense.basePathForCode": "/var/www/html",
    "laravel-pint.executablePath": "vendor/bin/pint",
    "laravel-pint.fallbackToGlobalBin": false,
}
``` 

It's kind of overkill, but no other way around it, as far as I can tell if you're using multiple extensions like me.

**Comment by szilvesztercsab** [comment]
A combination of this fake `php` executable:

```bash
#!/usr/bin/env bash

if [[ $1 == *"vendor"* ]]; then
    path=$1
    path=${path#*/vendor}
    path="/var/www/html/vendor$path"
    command="ddev php $path "${@:2}""
else
    command="ddev php "$@""
fi
$command
```

And this `settings.json`:

```json
{
    "php.debug.executablePath": ".vscode/php",
    "php.validate.executablePath": ".vscode/php",
    "Laravel.phpCommand": ".vscode/php \"{code}\"",
    "LaravelExtraIntellisense.phpCommand": "ddev php -r \"{code}\"",
    "LaravelExtraIntellisense.basePathForCode": "/var/www/html",
}
```

Seems to be the only working setup if you're using multiple extensions, like me.

**Comment by damms005** [comment]
Perhaps full support for DDEV environments will help with this? See #398

**Comment by TonsiTT** [comment]
This is fixed now thanks to @damms005 . I think this can be closed now
